Add cube background model class

mapazarr opened the pull request:
This PR is a continuation of PR #292.
Task list:

 - [x] the write method (and a test for it)
 - [x] porting the test file to `~gammapy-extra`
 - [x] make the test that use external files work
 - [x] implement `read_image` and `write_image` with a `WCS` object and a HDU table for the energy bin edges; tests
 - [x] separate the plot functions into `do_one_plot` and `do_many_plots` functions
 - [x] clean the code and autopep it
 - [x] the sphinx doc; add link to `examples/plot_background_model.py`; add examples of many plots (mosaic of images, stack of spectra)
 - [x] code review and final revision of inline comments

Further methods like histograming of event lists or interpolation and smoothing of the model will come on a later PR.

Sorry for the mess, I renamed the branch, which automatically closes the pull request. I thought the PR would be transparent to the rename.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412601
Ok, I moved the `WCS` functions to `gammapy.utils.wcs`. I named it `wcs` instead of the suggested `wcsutils` because being it already in `utils` makes the name redundant. In addition, none of the modules in `utils` are called `somethingutils`.

mapazarr commented:
I stopped the travis-ci because I found a problem on the WCS object. It seems that the header keywords exported to the header depend on how the wcs object is defined (as `~astropy.wcs` or `~astropy.wcs.WCS`. Or maybe I have a bug somewhere. I'll work on this later and rebase the commit history when I find a fix.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412707
You're right: the data members are already `Quantity` (or `Angle`) objects. But here's the thing.
This is the trick for storing the data in the FITS table: I need an array of an array, so that the table has only one line. I can either use `Quantity`, or `np.array`. The latter won't store the units, so I'd like to keep the `Quantity`, I think it's the most elegant and simple solution in order to store the data in the desired format, with the units that are already in the data members. Otherwise I'll need the lines that you wisely detected as not needed here:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412695

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412727
Yes, the bg models (in FITS files) from @mimayer do have a `E_THRES` header keyword defined. In his scripts, he defines it as the energy of the lowest boundary of the 1st energy bin (as I defined them). I followed his workflow there. I don't think an extra data member is needed for it, but I don't have a strong preference: if you still want to have it, please let me know.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412788
Ok, the `update` method worked! Thanks for the tip.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412799
OK, but it might get confusing if we call both formats (astropy table and HDU bin table) the same (`table`).

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412865
I have to strip the units and put them back on later: `np.log10` won't accept a quantity with units.
I leave it as it is, and put a TODO, that when PR #290 is done, it should be reviewed.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412897
https://github.com/gammapy/gammapy/pull/292#discussion_r34412944

Ok, we discussed this issue already here:

https://github.com/gammapy/gammapy/pull/292#discussion_r33191518
https://github.com/gammapy/gammapy/pull/292#discussion_r33191947

and I already simplified the methods. They are still long because I add 2 checks for a valid entry from the user. The method itself to find the bin is 4 lines on each case (X, Y and E). I'm not sure I can simplify it more.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34412967
OK, I removed the print statement from here (the message can be added to the exception, as you say).

You say it is no good practice to add print statements to the library code. I think sometimes it can be useful to have some printouts, especially if the code takes long to run (which is not the case), but also to get some important information. An example would be:
https://github.com/gammapy/gammapy/pull/292/files#diff-82cdfbacfcc0119d2ccf6621da08de7bR719
The user requests the plot for a particular energy, and it gets to know the boundaries of the slice used for the plot. This is maybe not the most useful case, but serves as an example for my question: how would you issue such an information?

I removed in any case all `print` statements of the class.

mapazarr commented:
About the plots:

https://github.com/gammapy/gammapy/pull/292#discussion_r34412977
I do need it for the color axis. So it stays.

https://github.com/gammapy/gammapy/pull/292#discussion_r34413000
https://github.com/gammapy/gammapy/pull/292#discussion_r34413005
I removed the save options.

https://github.com/gammapy/gammapy/pull/292#discussion_r34413022
OK, I was using for instance the `image` object to get the data in the tests of the plotting functions. But I learnt how to get them from the `ax` object, so no need for them. Now the functions return only the `ax` object.

mapazarr commented:
https://github.com/gammapy/gammapy/pull/292#discussion_r34413050
I know. Ok, I adopted your naming, though I find more explicative the old denomination of the functions.

mapazarr commented:
OK I just pushed a commit addressing all open comments in the old PR #292. I commented above on some of them (the non-trivial ones). Please have a look at my replies, and if applicable please answer to them on a new comment.

For future comments, please use this PR (#299).

mapazarr commented:
Permanent link to the docs:
https://googledrive.com/host/0BzXHZQx0oCLBfmcyUDdxMU9sLVduMXM1QXk5RTNYU3dXYmthVG9OZ18wWE1uR3lYaEsxRU0/html_doc/docs/_build/html

cdeil commented:
@mapazarr – I'm short on time this week, so I'll just do a final review tomorrow during the hangout.

Please update the task list and ask if you have any specific questions before you can finish this up.

Concerning the `E_THRES` header keyword ... if it's set to the energy array lower edge it's duplicated info and useless. But maybe in the future we do want to set it to indicate a valid range, or Gammalib requires it, so OK to keep if you like.

mapazarr commented:
Ok, I'm pretty much done with this PR. I think I have still one or 2 minor TODO's for you to review/comment on, on top of the questions I posted acouple of days ago above, and of course the final code review.
The link to the docs is here:
https://googledrive.com/host/0BzXHZQx0oCLBfmcyUDdxMU9sLVduMXM1QXk5RTNYU3dXYmthVG9OZ18wWE1uR3lYaEsxRU0/html_doc/docs/_build/html

cdeil commented:
This example doesn't work for me:
```
$ python examples/plot_background_model.py 
Traceback (most recent call last):
  File "examples/plot_background_model.py", line 23, in <module>
    write_kwargs=dict(clobber=True))
  File "/Users/deil/code/gammapy/gammapy/background/models.py", line 475, in write
    self.to_fits_image().writeto(outfile, **write_kwargs)
  File "/Users/deil/code/gammapy/gammapy/background/models.py", line 432, in to_fits_image
    imhdu.header.update(wcs_header)
  File "/Users/deil/Library/Python/3.4/lib/python/site-packages/astropy-1.1.dev12994-py3.4-macosx-10.10-x86_64.egg/astropy/io/fits/header.py", line 1167, in update
    update_from_dict(k, other[k])
  File "/Users/deil/Library/Python/3.4/lib/python/site-packages/astropy-1.1.dev12994-py3.4-macosx-10.10-x86_64.egg/astropy/io/fits/header.py", line 1149, in update_from_dict
    card = Card(k, v)
  File "/Users/deil/Library/Python/3.4/lib/python/site-packages/astropy-1.1.dev12994-py3.4-macosx-10.10-x86_64.egg/astropy/io/fits/card.py", line 448, in __init__
    self.value = value
  File "/Users/deil/Library/Python/3.4/lib/python/site-packages/astropy-1.1.dev12994-py3.4-macosx-10.10-x86_64.egg/astropy/io/fits/card.py", line 570, in value
    raise ValueError('Illegal value: %r.' % value)
ValueError: Illegal value: .
```
The problem are the `''` and `'COMMENT'` entries:
```
In [1]: dict(wcs_header)
Out[1]: 
{'': ,
 'CDELT1': 0.546914274042,
 'CDELT2': 0.546914274042,
 'COMMENT':  WCS header keyrecords produced by WCSLIB 5.8,
 'CRPIX1': 0.5,
 'CRPIX2': 0.5,
 'CRVAL1': -6.01605701447,
 'CRVAL2': -6.01605701447,
 'CTYPE1': 'DETX',
 'CTYPE2': 'DETY',
 'CUNIT1': 'deg',
 'CUNIT2': 'deg',
 'LATPOLE': 90.0,
 'WCSAXES': 2}
```

Merging the headers like this works for me, so that's what I'd suggest as a workaround:
```python
header = wcs_header.update(imhdu.header)
```
The reason this error doesn't show up for you locally and on travis-ci is probably that you're not using the very latest Astropy master.
I'll make an issue that travis-ci docs builds with Python 3 and with the latest astropy master should be added.

cdeil commented on docs/background/index.rst:
```diff
@@ -27,11 +27,53 @@ TODO
 
+.. _bg_models:
+
+Background Models
```
I'd like the top-level `gammapy.background` page to remain short, just giving an overview of the available functionality and maybe one or two quick code examples / plots.

So please move this whole section to a sub page.
An example how to do this is here: https://gammapy.readthedocs.org/en/latest/image/plotting.html , i.e. plotting is a subpage of the image docs.

cdeil replied:
PS: the docs you put online on Google drive don't work (I get 404 page not found errors when I click on gammapy.background).
Contrary to what I said before: please don't put docs online any more! I'll simply check out your branch, built the docs locally and make inline comments on what I think should be fixed, using screenshots where needed. This works nicely.

mapazarr replied:
Done.

cdeil commented on docs/background/index.rst:
```diff
@@ -27,11 +27,53 @@ TODO
+The file comes originally from the `~GammaLib` repository but has
+been slightly modified.
+
+TODO: the scripts produce some white canvases that I need to remove!
```
Resolve this TODO.

mapazarr replied:
It was resolved already. I forgot to remove the TODO.
Done.

cdeil commented on examples/plot_background_model.py:
```diff
@@ -1,99 +1,25 @@
+from gammapy import datasets
+
+# read
+filename = '../test_datasets/background/bg_cube_model_test.fits'
```
Remove this line.
Also the `#read`, `#plot`, `#write` comment lines can be removed IMO ... you've implemented a beautiful, clean API and there's no need to repeat in comments what the code is clearly expressing. :-)

mapazarr replied:
Fixed.

cdeil commented on docs/background/index.rst:
```diff
@@ -27,11 +27,53 @@ TODO
+   :include-source:
+
+More complex plots can be easily produced with a few lines of code:
+TODO: mosaic/stack plots examples!!!
```
Resolve this TODO.

I don't see the need for three plotting scripts, my suggestion would be to:
- Remove the plotting from the first code example ... just have that read and write.
- Combine the other two plotting scripts in one. Show three panels in one row (using `plt.subplots`), which show two images (one at very low and one at very high energy, so that they look different), and for the spectrum loop over a handful positions and draw the curves in the third panel.

cdeil replied:
I currently get this:

![image](https://cloud.githubusercontent.com/assets/852409/8694818/dbf53f7e-2ae1-11e5-9f4e-4a67e7d9194f.png)

Why isn't the background rate a power-law, with lower rate at higher energy?

mapazarr replied:
The TODO was resolved already. I forgot to remove it.
Done.

Changed the plot scheme to have only 1 plot with 3 pads.

The plots are correct. I still need to change the test file to have a more meaningful example.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+            table containing the bg cube
+        """
+        # data arrays
+        a_detx_lo = Quantity([self.detx_bins[:-1]])
```
Sorry if this has been discussed already, I don't remember: Why call `Quantity` here? `self.detx_bins` is already a quantity, so this is useless, no?

mapazarr replied:
See this: https://github.com/gammapy/gammapy/pull/299#issuecomment-120673603

mapazarr replied:
I can add an explanation in the code as a comment if you like.

cdeil replied:
So you basically want to reshape the array here?
Then please call reshape or maybe [numpy.expand_dims](http://docs.scipy.org/doc/numpy/reference/generated/numpy.expand_dims.html) to make the code clearer.
(I hope it works with Quantity objects).

Also, in this case the temp variables aren't really needed, you can just assign to the table columns below and it'll still nicely fit on one line, no?

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+
+        table.meta['E_THRES'] = a_energy_lo.flatten()[0].value
+
+        # name
```
Remove this comment.

mapazarr replied:
Fixed.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+        tbhdu : `~astropy.io.fits.BinTableHDU`
+            table containing the bg cube
+        """
+        # build astropy table and convert it to fits bin table
```
Remove this comment ... it says exactly the same thing as the code on the next line.

mapazarr replied:
Fixed.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+
+        enhdu = _table_to_fits_bin_table(energy_table)
+
+        hdu_list = fits.HDUList([fits.PrimaryHDU(), imhdu, enhdu])
```
Why put an empty `PrimaryHDU`?
Please make a `PrimaryHDU` directly instead of an `ImageHDU`, and then there's no need for an empty `PrimaryHDU`, right?

mapazarr replied:
The issue with the primary HDU is that it's not possible to give it a name (in this case 'BACKGROUND'), which makes it a bit less easy to identify. I have seen other FITS images which use an empty Primary HDU, so I followed the trend. I don't have a strong opinion on this, but if you also don't, I'd like to keep it as it is now. Please let me know if you still want me to change it.

cdeil replied:
As I said, I prefer to have the cube in the primary HDU.
Please change this here for the background cube and I think just fixing the SpectralCube write method in a commit here is better, no need for a separate PR.

mapazarr replied:
OK. Done. It was also a bit of work to make the `SpectralCube.writeto` method to work.
Please see the TODO that I left inside.
This solves issue #304 .

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+
+        return hdu_list
+
+    def write(self, outfile, format='table', write_kwargs=None):
```
I'd prefer you use `**kwargs` here instead of `write_kwargs=None`.
That's the common pattern and it should work.

I remember this has been discussed before, and there was some issue.
But please, let's try to make this work and re-discuss if needed.
One thing you have to know is that `**kwargs` always has to be at the end.

mapazarr replied:
Done, though I have seen the other pattern in other functions as well.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+        """
+        detx_edges_low = self.detx_bins[:-1]
+        detx_edges_high = self.detx_bins[1:]
+        detx_edges_centers = (detx_edges_low + detx_edges_high)/2.
```
Does this give the same result?
```python
detx_centers = self.detx_bins[:-1] + np.diff(self.detx_bins)
```

mapazarr replied:
Almost: you need to divide diff by 2 :-)
Done.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+        energy_edges_high = self.energy_bins[1:]
+        e_lo_tev = energy_edges_low.to('TeV').value
+        e_hi_tev = energy_edges_high.to('TeV').value
+        energy_bin_centers = 10.**(0.5*(np.log10(e_lo_tev*e_hi_tev)))
```
Does this give the same result?
```python
log_bin_edges = np.log(self.energy_bins.value)
log_bin_centers = log_bin_edges[:-1] + np.diff(log_bin_edges)
bin_centers = Quantity(np.exp(log_bin_centers), self.energy_bins.unit)
```

mapazarr replied:
You are missing again a division by 2 in the diff.
Both methods differ by at most 1e-5 in some values.
Done.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+        from matplotlib.colors import LogNorm
+        from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+        energy = energy.flatten() # flatten
```
Remove comment `# flatten` ... it doesn't add info to what the code says.

mapazarr replied:
Done.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+                                                      energy_bin_edges.unit)
+
+        # get data for the plot
+        ii = energy_bin
```
There's no point in creating a second variable `ii` that is the same as `energy_bin`, right?
Just use `energy_bin`?

mapazarr replied:
Done.

cdeil commented on gammapy/background/tests/test_models.py:
```diff
@@ -43,3 +49,100 @@ def test_y_model(self):
+    def test_read_bin_table(self):
+
+        # test shape of bg cube when reading a file
+        filename = '../test_datasets/background/bg_cube_model_test.fits'
```
remove this line

mapazarr replied:
Done overall. Please notice, that in this case, the `get_path` line is too long (beyond 80 chars).

cdeil replied:
Looks OK to me no, no worries if you go > 80 characters in some cases that would be awkward to wrap, like filenames or URLs. My point about insisting on line length limit sometimes is that code should be readable, which it isn't if one has to scroll back and forth to read a very long expression.

cdeil commented on gammapy/background/tests/test_models.py:
```diff
@@ -43,3 +49,100 @@ def test_y_model(self):
+    @remote_data
+    def test_image_plot(self):
+
+        filename = '../test_datasets/background/bg_cube_model_test.fits'
```
remove this line

mapazarr replied:
Done.

cdeil commented on gammapy/background/tests/test_models.py:
```diff
@@ -43,3 +49,100 @@ def test_y_model(self):
+    @remote_data
+    def test_spectrum_plot(self):
+
+        filename = '../test_datasets/background/bg_cube_model_test.fits'
```
remove this line

mapazarr replied:
Done.

cdeil commented on gammapy/background/tests/test_models.py:
```diff
@@ -43,3 +49,100 @@ def test_y_model(self):
+    @remote_data
+    def test_write_bin_table(self):
+
+        filename = '../test_datasets/background/bg_cube_model_test.fits'
```
remove this line

mapazarr replied:
Done.

cdeil commented on gammapy/background/tests/test_models.py:
```diff
@@ -43,3 +49,100 @@ def test_y_model(self):
+    @remote_data
+    def test_read_write_image(self):
+        
+        filename = '../test_datasets/background/bg_cube_model_test.fits'
```
remove line.

mapazarr replied:
Done.

cdeil commented:
Add a Sphinx automodapi directive for `gammapy.utils.wcs`, so that the API docs for that new sub-module show up in the docs:
https://github.com/gammapy/gammapy/blame/master/docs/utils/index.rst#L27

cdeil commented:
@mapazarr – Did another round of review. Nothing major, but lots of little things for you to address. Let's see if this is ready to be merged tonight ...

cdeil commented:
One more thing: please add a changelog entry matching the title of this PR (which I just now changed).

cdeil commented:
If you have time, check that the code you've written is covered by a test:
```
time python setup.py test -P background --coverage --remote-data
open htmlcov/gammapy_background_models.html
``

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+                                     bin_edges_y=self.detx_bins)
+        return wcs
+
+    def find_det_bin(self, det):
```
If you want to keep this, then split this:
```python
get_det_bin_index
get_det_bin_edges (calls get_det_bin_index)
maybe even get_det_bin_value
```

mapazarr replied:
Done. I didn't add a `get_det_bin_value` yet, since it's not needed at the moment.

mapazarr commented:
OK, I just updated my astropy version to HEAD and it doesn't work??!
I used
```
git fetch astropy master
git rebase astropy/master
```
Then I run:
`python setup.py develop`
and I get:
```
Traceback (most recent call last):
  File "setup.py", line 58, in <module>
    package_info = get_package_info()
  File "/home/mapaz/astropy/development_code/astropy/astropy_helpers/astropy_helpers/setup_helpers.py", line 539, in get_package_info
    ext_modules.extend(setuppkg.get_extensions())
  File "astropy/wcs/setup_package.py", line 194, in get_extensions
    generate_c_docstrings()
  File "astropy/wcs/setup_package.py", line 127, in generate_c_docstrings
    from astropy.wcs import docstrings
  File "/home/mapaz/astropy/development_code/astropy/astropy/wcs/__init__.py", line 30, in <module>
    from .wcs import *
  File "/home/mapaz/astropy/development_code/astropy/astropy/wcs/wcs.py", line 57, in <module>
    _parsed_version = _wcs.__version__.split('.')
AttributeError: 'module' object has no attribute '__version__'
```
Is there a bug/problem in astropy? It seems that there is something missing.

mapazarr commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,724 @@ def evaluate(self, x, y):
+        wcs_header = self.det_wcs.to_header()
+
+        # transferring header values
+        imhdu.header.update(wcs_header)
```
According to your comment
https://github.com/gammapy/gammapy/pull/299#issuecomment-121530634

I tried to update my astropy-dev to the HEAD version but it fails, so I can't reproduce your error.

I can say, that I don't see these 2 entries in the `''` and `'COMMENT'` in the wcs header, using the astropy stable version. It looks like it's a problem in astropy, rather than in my code?

About your suggested workaround: as I can't reproduce the error, I can't test it. But 2 thoughts:

1. It looks a bit cumbersome: update the wcs header with the image header, then overwrite the image header.

2. I don't see how this procedure should get rid of the problematic entries, since they should remain in the end version of the header, am I right?

So how should we proceed? File an issue to astropy about the header problem? (and maybe ask for the problem fetching the HEAD version?)

cdeil replied:
I don't understand why you didn't just apply the fix I suggested yesterday.
It's what it took to make this work for me, and you can test that it still works for you and on travis-ci.
Now with this known issue I still can't re-test and merge this.
Please apply the suggested change.
If there's still some issue we'll think of a different fix, but I think it'll be OK.

mapazarr replied:
OK, I cleaned the astropy-dev version and it works now. Thanks for the tip.
As for your fix, it doesn't work:
Your suggestion:
`header = wcs_header.update(imhdu.header)`
leaves the header empty, since the correct line would be
`imhdu.header = wcs_header.update(imhdu.header)`
But then I get the following error:
```
Traceback (most recent call last):
  File "plot_background_model.py", line 18, in <module>
    bg_model.write('{}_image.fits'.format(outname), format='image', clobber=True)
  File "/home/mapaz/astropy/development_code/gammapy/gammapy/background/models.py", line 473, in write
    self.to_fits_image().writeto(outfile, **kwargs)
  File "/home/mapaz/astropy/development_code/gammapy/gammapy/background/models.py", line 432, in to_fits_image
    imhdu.header = wcs_header.update(imhdu.header)
  File "/home/mapaz/astropy/development_code/astropy/astropy/io/fits/hdu/image.py", line 188, in header
    self.update_header()
  File "/home/mapaz/astropy/development_code/astropy/astropy/io/fits/hdu/image.py", line 931, in update_header
    super(PrimaryHDU, self).update_header()
  File "/home/mapaz/astropy/development_code/astropy/astropy/io/fits/hdu/image.py", line 264, in update_header
    old_naxis = self._header.get('NAXIS', 0)
AttributeError: 'NoneType' object has no attribute 'get'
```
So we need to think about how to fix this. Maybe it's worth to post an issue in Astropy?

mapazarr replied:
Maybe a (temporary) workaround would be either:
* copy the header kaeywords 1 by 1, as I was doing before
* remove the problematic entries 1 by 1 before updating the header.
What do you think?

cdeil replied:
Well, you or I need to learn how `header.update` works, it's not clear to me is this is a bug or something that can be improved in Astropy, or just a lack of understanding on our side.
So what should be done is to reproduce the issue with a minimal test case that's just a few lines of Astropy code, not connected to Gammapy. Only then does it make sense to post an issue or ask for help on astropy-dev if there's something about it we don't understand.
But this takes time ... maybe an hour, but easily half a day.
OK if you want to work on this now, but just not writing a WCS to file (commenting this out) and leaving a TODO and moving on quickly is also OK ... your choice.

cdeil replied:
Only implement workarounds if it's quick (10 minutes) to do and leave a TODO, i.e. don't spend much time on something that will very likely be fixed in another way in the future.

mapazarr replied:
Calling the wcs header directly when creating the HDU works:
```
        imhdu = fits.PrimaryHDU(data=self.background.value,
                                header=self.det_wcs.to_header())
```
So, fixed.

cdeil replied:
:tada:

mapazarr commented:
@cdeil 
About: https://github.com/gammapy/gammapy/pull/299#issuecomment-121561073
I already added the lines
```
.. automodapi:: gammapy.utils.wcs
    :no-inheritance-diagram:
```
to the file.

cdeil commented:
About your `AttributeError` for `__version__` in `astropy.wcs`: I don't know.
A quick search of the Astropy issue tracker doesn't show a mention of this, and astropy master works for me and many other developers, so it's most likely a setup issue on your machine.

One thing you could try it running `python setup.py build` in a fresh git clone.
Or use `git clean -fdx` to get rid of old files in your existing git repo.
It might be old or outdated files lying around and `astropy.wcs` changed from wcslib 4 to 5 and I think also changed the build system a bit in the past months.

mapazarr commented:
@cdeil 
Your fix doesn't work (see details above). It was working for the example because it didn't need to access the header explicitly. But the tests fail.
I'm going to bed now and continue tomorrow. Maybe you come with a new idea on the meantime?

cdeil commented:
> Maybe you come with a new idea on the meantime?

I won't this week, no time, sorry.
Good night!

Leave a comment when I should merge or review this PR again tomorrow.

mapazarr commented:
@cdeil Ok, I implemented all comments. There is still a couple of small TODO's in the code. Please revise and let me know your thoughts.

I'm not sure which WARNING is causing the travi-ci sphinx docs to fail. I think it's this one:
```
WARNING: intersphinx inventory 'http://scikit-learn.org/stable/objects.inv' not fetchable due to <class 'socket.error'>: [Errno 104] Connection reset by peer
```
(it looks like the WEbsite ist nicht verfuegbar):
```
The sourceforge.net website is temporarily in static offline mode.
Only a very limited set of project pages are available until the main website returns to service.
```
Here is the new "permanent" link to the docs:
https://ms2.physik.hu-berlin.de/~mapaz/gammapy/docs/_build/html/

cdeil commented on CHANGES.rst:
```diff
@@ -51,6 +51,8 @@ Pull requests
 - Function to plot Fermi 3FGL light curves [#286] (Jonathan Harris)
 - Add EffectiveAreaTable exporter to EffectiveAreaTable2D [#276] (Johannes King)
 - Add interface to HESS FitSpectrum JSON output [#296] (Christoph Deil)
+- Remove gammapy.shower package [#291] (Christoph Deil)
```
I don't understand why the `gammapy.shower` line shows up in the diff for this PR!?
I added it here: https://github.com/gammapy/gammapy/pull/291/files
Now in master it doesn't seem to be there any more: https://github.com/gammapy/gammapy/blame/master/CHANGES.rst#L51

Well ... never mind ... if it's not there it's good it's being added back.

mapazarr replied:
OK. I leave it as it is now.

cdeil commented on docs/background/index.rst:
```diff
@@ -20,6 +20,8 @@ and Section 7.4 "Acceptance determination and predicted background"
 in [Naurois2012]_ describe mostly the same methods as [Berge2007]_,
 except for the "2D acceptance model" described in Section 7.4.3.
 
+The backgrund models implemented in Gammapy are documented in :ref:`bg_models`.
```
Typo: `backgrund` -> `background`

mapazarr replied:
Fixed.

cdeil commented on docs/background/plot_bgcube.py:
```diff
@@ -0,0 +1,30 @@
+fig.set_size_inches(30., 8., forward=True)
+
+# plot images
+axes[0] = bg_model.plot_image(energy=Quantity(0.5, 'TeV'), ax=axes[0])
```
You can remove the `axes[0] = ` part, it has no effect (same in the calls below).

mapazarr replied:
Fixed.

cdeil commented on docs/background/plot_bgcube.py:
```diff
@@ -0,0 +1,30 @@
+axes[0] = bg_model.plot_image(energy=Quantity(0.5, 'TeV'), ax=axes[0])
+axes[1] = bg_model.plot_image(energy=Quantity(50., 'TeV'), ax=axes[1])
+
+# plot spectra
```
The example file still looks weird (background rate increasing with energy).
Can you change to the other one we talked about?

mapazarr replied:
Yes I know. I have implemented a function to create test cube bg models. I am adding this to this PR. Please see this PR in `~gammapy-extra`: https://github.com/gammapy/gammapy-extra/pull/18

cdeil commented on docs/development/index.rst:
```diff
@@ -416,3 +415,30 @@ A lot of code that's not gamma-ray specific belongs in other packages
+use the `.value` methods of these object types. Don't forget to
+remove the call to `.value` after performing the test.
+
+* TODO: should we submit a bug report to Astropy?
```
Is this about the `repr failed` issue? I think we said that yes, this is probably a bug and worth filing an Astropy issue?

In any case, I'd suggest to remove the `Tip` and `TODO` here and instead point to this utility function:
https://gammapy.readthedocs.org/en/latest/api/gammapy.utils.testing.assert_quantity.html
(and if you're in the mood, add an example to that docstring how to use it).

mapazarr replied:
Ok I filed the issue: https://github.com/astropy/astropy/issues/3978
and replaced the `Tip` and `TODO` by the reference. I didn't add an example, tough I added an equivalent method for `Angle`.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +82,740 @@ def evaluate(self, x, y):
+
+        Parameters
+        ----------
+        outfile : `~str`
```
For Python built-in types, let's not create intersphinx links everywhere, i.e. just type:
```
outfile : str
    ...
```
Please change this everywhere in this PR

mapazarr replied:
OK, fixed. Though I think it could be convenient sometimes in order to quickly check for a list of methods, and I don't think it harms to have these links.

cdeil commented on gammapy/data/spectral_cube.py:
```diff
@@ -413,29 +413,43 @@ def to_fits(self):
+        image.header['SPEC_UNIT'] = '{0.unit:FITS}'.format(self.data)
+
+        # for BinTableHDU's the data must be added via a Table object
+        # TODO: this could be simplified a bit by using
```
You decide, but please resolve this TODO.
You can also just import the private utility function from where it is now if it's useful.

mapazarr replied:
Fixed. I moved the method to `~gammapy.utils.fits.table_to_fits_bin_table` and made it public.

cdeil commented on gammapy/data/spectral_cube.py:
```diff
@@ -413,29 +413,43 @@ def to_fits(self):
             Name of output file (.fits)
-        overwrite : bool
-            Overwrite existing output file?
+        ###overwrite : bool
```
How do you want to handle the `overwrite`?
I guess removing it is fine?
Or keep it, document it and convert it to clobber?
You decide, but please finish this up.

mapazarr replied:
Sorry I forgot to remove the comments. I thought to handle it as for `CubeBackgroundModel`: pass all extra arguments (including `clobber`) to the corresponding FITS I/O `writeto` method. Fixed.

cdeil commented:
This is very close ... I left some more inline comments ... I'll merge it tomorrow.
I don't know if the failing intersphinx download caused the Sphinx build error, but I don't see any other new warnings there ... let's just see if it goes away.

mapazarr commented:
Ok, I implemented all the comments. I also added a method to create test `CubeBackgroundModel` objects. It helped for definitely solving the uncertainty in the axis order for the bacground cube data: `(E, y, x)`. I am ready for merging this and start to work on cleanup PRs for the 0.3 release tomorrow.

mapazarr commented:
About the problems in travis-ci:

* The spinx doc fails. I fixed the warnings related to the code/docs I changed, but it still fails. I get among others, this warning:
  `WARNING: intersphinx inventory 'http://scikit-learn.org/stable/objects.inv' not fetchable due to <class 'urllib2.HTTPError'>: HTTP Error 503: Service Unavailable`
  So I don't think it is related to my modes.

* It fails to import `matplotlib.pyplot` in the plotting functions, when not installing optional dependencies. I guess `matplotlib` is not required for `gammapy`. So, should I add special decorators to the functions using them? (please add link to example).

mapazarr commented:
@cdeil Please notice the issue #305, that I'm experiencing, when reviewing the docs for this PR.

mapazarr commented on gammapy/background/models.py:
```diff
@@ -71,3 +83,699 @@ def evaluate(self, x, y):
+        det_bin = bg_cube_model.find_det_bin(det=Angle([0., 0.], 'degree'))
+        bg_cube_model.background[energy_bin, det_bin[1], det_bin[0]]
+    """
+    # TODO: how can I link to this example in the models.rst doc subpage?!!!
```
@cdeil PLease consider this TODO before merging. Please give me some input.

cdeil replied:
You have to define a Sphinx label and then add a Sphinx reference (like in LaTeX).

Example: https://github.com/gammapy/gammapy/blob/e781e88de4a55f0f20473b062ea6eb38dd51f3ed/gammapy/obs/observation.py#L45

This should work the same from the high-level docs or the docstrings.

mapazarr replied:
Thanks for the suggestion. I was actually thinking of having the code example from the stringdpcs show in the high-level docs, without the need to repeat the code-block, not just a link. Is it possible? (how?)

cdeil replied:
The only way to achieve this would be a `..literalinclude` of an extra `.py` file from the docstring and the high-level docs. I've never seen anyone use `..literalinclude` from a docstring, so I'm not sure if it works. And anyways it's not a good solution because in interactive use from IPython people wouldn't see the code example.

So you'll have to pick: duplicate the code example or add a sentence linking to it in another location. I think both are OK.

mapazarr replied:
OK, yes. It sounds a bit complicated. I'm duplicating the code block then.

cdeil commented on gammapy/background/tests/test_models.py:
```diff
@@ -43,3 +50,97 @@ def test_y_model(self):
+                                                  len(bg_cube_model.dety_bins) - 1,
+                                                  len(bg_cube_model.detx_bins) - 1)
+
+    def test_image_plot(self):
```
To get rid of this fail:
https://travis-ci.org/gammapy/gammapy/jobs/71722967#L991

you have to add a decorator as in this example:
https://github.com/gammapy/gammapy/blob/9c979c56ed278b7deda1cada1788241be573e577/gammapy/time/tests/test_plot.py#L15

mapazarr replied:
Fixed. Thanks for the example.

cdeil commented on docs/background/models.rst:
```diff
@@ -0,0 +1,51 @@
+   det_bin = bg_cube_model.find_det_bin(det=Angle([0., 0.], 'degree'))
+   bg_cube_model.background[energy_bin, det_bin[1], det_bin[0]]
+
+TODO: how can I link here the exmaple from the CubeBackgroundModel cass?!!! (to avoid duplication)!!!
```
Maybe just write this?
```
See the example in the `~gammapy.background.CubeBackgroundModel` class docstring.
```

mapazarr replied:
OK, I'm using this for now. But is it possible to show/link the example here without repeating the code-block?

cdeil commented on docs/development/index.rst:
```diff
@@ -416,3 +415,28 @@ A lot of code that's not gamma-ray specific belongs in other packages
+dozens of other available asserts from pytest or numpy in various
+places.
+
+More details in `numpy.testing.assert_allclose`.
```
Remove that line (that's a given that the numpy docs contain more info, that's why you link to it above).

mapazarr replied:
OK. Done. About the `~gammapy.utils.testing.assert_quantity` method. Please consider issue #306. Maybe we can discuss/address this on a separate PR?

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -209,3 +211,128 @@ def make_test_observation_table(observatory_name, n_obs, debug=False):
     return obs_table
+
+
+def make_test_bg_cube_model(detx_min=Angle(-10., 'degree'),
```
Maybe use `detx_range`, `dety_range` and `energy_band` as parameters to have fewer parameters (but same functionality)?

`index` -> `spectral_index`

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -209,3 +211,128 @@ def make_test_observation_table(observatory_name, n_obs, debug=False):
+    powerlaw = (energy_points/E_0)**-index
+    background = norm*gaussian*powerlaw
+
+    # apply mask if requested
```
IMO numpy masked arrays cause trouble (performance issues, unexpected behaviour with a few numpy functions) sooner or later.
My suggestion would be to filling zeros here instead of making a masked array!?

mapazarr replied:
It was working fine (there is a test testing the masked option). I anyway changed it in order not to use a masked array. I define the mask and multiply by the bg array. So fixed.

cdeil commented on gammapy/utils/wcs.py:
```diff
@@ -0,0 +1,133 @@
+
+    Parameters
+    ----------
+    w : `~astropy.wcs.WCS`
```
Rename `w` -> `wcs`

Also, I usually prefer `to` methods for converters, i.e. this should be `linear_wcs_to_arrays`.
I know this is arbitrary, but would you mind changing?

mapazarr replied:
Ok. I was using `w`, because I have to call the `wcs` method, so `wcs.wcs` seemed too much. But I did as you suggested.

I updated the method names, for the second time :-)

cdeil commented on gammapy/utils/wcs.py:
```diff
@@ -0,0 +1,133 @@
+
+    Returns
+    -------
+    w : `~astropy.wcs.WCS`
```
Please call this return parameter `wcs` and also the object in the code.

mapazarr replied:
Done.

cdeil commented:
I left a few inline comments. I don't have much time today, maybe I can have another look late tonight.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +83,699 @@ def evaluate(self, x, y):
+
+        # get energy values as a table HDU, via an astropy table
+        energy_table = Table()
+        energy_table['ENERGY_BIN_EDGES'] = self.energy_bins
```
I think the convention from X-ray and Fermi is to call the extension `EBOUNDS` and the column `ENERGY`.
Maybe also leave a TODO that this should be used soon:
https://github.com/gammapy/gammapy/pull/290/files#diff-02eca90c3140a78dc540f64e561a7592R190

mapazarr replied:
Done.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +83,699 @@ def evaluate(self, x, y):
+        Several output formats are accepted, depending on the value
+        of the `format` parameter:
+
+        * bin_table (default and preferred format): `~astropy.io.fits.BinTableHDU`
```
`bin_table` -> `table`

mapazarr replied:
Done.

cdeil commented on gammapy/background/models.py:
```diff
@@ -71,3 +83,699 @@ def evaluate(self, x, y):
+        dety_edges_centers : `~astropy.coordinates.Angle`
+            array of image bin centers (Y coord)
+        """
+        detx_edges_centers = self.detx_bins[:-1] + np.diff(self.detx_bins)/2.
```
Thinking about this some more, I think this is a tick more readable:
```python
bin_centers = 0.5 * (bin_edges[:-1] + bin_edges[-1:])
```
and is more similar to the expression for log bin centers (geometric mean):
```
bin_centers = np.sqrt(bin_edges[:-1] * bin_edges[1:])
```
So please use that.

It's also faster! :speedboat: :racehorse: 
:-)

mapazarr replied:
Done. It is how I had done it before, only I had temp vars for `bin_edges[:-1]` and `bin_edges[1:]` for readability purposes...

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -209,3 +211,128 @@ def make_test_observation_table(observatory_name, n_obs, debug=False):
+        detx_center = (detx_max + detx_min)/2.
+        dety_center = (dety_max + dety_min)/2.
+        mask = (detx_points > detx_center) & (dety_points > dety_center)
+        #import IPython; IPython.embed()
```
Remove these comments

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/tests/test_make.py:
```diff
@@ -43,3 +44,34 @@ def test_make_test_observation_table():
+    bg_cube_model = make_test_bg_cube_model(apply_mask=True)
+
+    # test that values with (x, y) > (0, 0) are zero
+    x_points = Angle(np.arange(5), 'degree') + Angle(0.01, 'degree')
```
Slightly shorter:
```python
x_points = Angle(np.arange(5) + 0.01, 'degree')
```

mapazarr replied:
Done.

cdeil commented on gammapy/utils/testing.py:
```diff
@@ -15,7 +17,21 @@ def assert_quantity(actual, desired, *args, **kwargs):
+    assert_allclose(actual, desired, *args, **kwargs)
+
+
+def assert_angle(actual, desired, *args, **kwargs):
```
I don't think this is really useful in addition to the quantity assert helper function?

mapazarr replied:
Ok. I removed it. The other one should fall soon also :-)

cdeil commented on gammapy/utils/wcs.py:
```diff
@@ -0,0 +1,133 @@
+
+    Parameters
+    ----------
+    name_x : `~string`
```
string -> str

mapazarr replied:
Done.

cdeil commented on gammapy/utils/wcs.py:
```diff
@@ -0,0 +1,133 @@
+    w.wcs.cunit = [unit_x, unit_y]
+    w.wcs.cdelt = [delta_x.to(unit_x).value, delta_y.to(unit_y).value]
+    # ref as lower left corner (start of (X, Y) bin coordinates)
+    # coordinate start empiricaly determined at pix = 0.5: why 0.5?
```
I think we discussed this and it's understood, so remove the "why 0.5?" comment?

mapazarr replied:
Sorry I forgot. Done.

cdeil commented on gammapy/utils/wcs.py:
```diff
@@ -0,0 +1,133 @@
+        raise ValueError("The Y bins are not linear! Diff = {}".format(np.diff(bin_edges_y.value)))
+
+    # Create a new WCS object. The number of axes must be set from the start
+    w = WCS(naxis=2)
```
`w` -> `wcs` (single-letter variables are a bit evil)

mapazarr replied:
I have changed it already. Maybe you're looking at an old version.

cdeil commented:
Left some more minor inline comments.

mapazarr commented:
Ok I implemented the last comments. I also fixed some problems in SpectralCube nad the tutorials, that where preventing the sphinx job in travis-ci to work.

Can we merge this PR already? Some of the minor comments are turning around on issues discussed a few times already, and undoing some of the proposed changes.

cdeil commented:
Merging this now.
Thanks for fixing all the issues that came up in this PR or following up in other Astropy / Gammapy issues!

I looked through the coverage report
```
time python setup.py test --remote -V --coverage
open htmlcov/gammapy_background_models.html
```
and generally coverage is good.
The only thing that's not tested is the code under `except ValueError` in `_parse_bg_units`
Why is that needed at all? Maybe because the background cubes from Michael or Gammalib don't have correct units per the FITS standard? If so we should tell them to fix it and make a note that this unit handling helper function should be removed as soon as it's done.

cdeil commented:
:fireworks:

cdeil commented:
Docs build on readthedocs worked: https://gammapy.readthedocs.org/en/latest/background/models.html

mapazarr commented:
Hi Christoph. Indeed the existence of `_parse_bg_units` is because of the units in @mimayer's files. I checked and they actually  comply the FITS standard, but they are not understood by `~astropy.units.Units`.

Please see the text at the examples section of: https://heasarc.gsfc.nasa.gov/docs/heasarc/ofwg/docs/general/ogip_93_001/ogip_93_001.html

'the number of spaces between individual unit sub-strings is left a matter for personal preference'

@mimayer doesn't use any spaces, i.e. `1/s/MeV/sr`, while `Quantity` expects spaces and can't parse this unit, i.e. `1 / s / MeV / sr`.

According to the FITS standards both are valid.

This could be another issue for astropy.

mapazarr commented:
I didn't get any coveralls report.

cdeil commented:
You have to manually go to https://coveralls.io/github/gammapy/gammapy (I always do it via the link [here](https://github.com/gammapy/gammapy#status-shields)) and find the right build.
There is the possibility to configure Coveralls to post comments on PRs with links to the right build.
But so far we weren't using that much and it was annoying.

mapazarr commented:
BTW, I think the test file from `gammalib` also was having the issue with the units.
